thorlcr/master/thmastermain.cpp

```diff
@@ -358,7 +358,39 @@ class CRegistryServer : public CSimpleInterface
                     EXCLOG(e, "Shutting down slave");
                     e->Release();
                 }
+                if (watchdog)
+                    watchdog->removeSlave(ep);
+            }
+        }
+
+        CTimeMon tm(20000);
+        unsigned numReplied = 0;
+        while (numReplied < slavesRegistered)
+        {
+            unsigned remaining;
+            if (tm.timedout(&remaining))
+            {
+                PROGLOG("Timeout waiting for Shutdown reply from slave(s) (%u replied out of %u total)", numReplied, slavesRegistered);
+                break;
+            }
+            try
+            {
+                rank_t sender;
+                CMessageBuffer msg;
+                if (queryNodeComm().recv(msg, RANK_ALL, shutdownTag, &sender, 50))
```

Comment:
But why have such a small timeout on recv, which could potentially be exceeded if either master was busy.

Why not instead rely on the signalling mechanism built into MP and let the queryNodeComm().recv() handle the timeout. I don't see the upside of handling like this with small timeout and looping.
Passing the overall timeout (20s) to recv would achieve same thing and would avoid the need for Sleep and avoid the potential for a 50ms timeout being missed by if network or machine was busy.